File: Chapter_3/number_theory.java
/*
* TODO-based exploration of number theory with SymPy.
* Author: ChatGPT with small adaptations by Miguel Vargas Martin
* Last modified: January 2026
*
* Instructions:
* - Read each TODO carefully.
* - Before running a line, try to predict its output.
* - Uncomment lines only when instructed.
* - Answer the questions in the comments or in a separate worksheet.
*/

import java.math.BigInteger;
import java.util.*;

import org.apache.commons.math3.util.ArithmeticUtils;
import org.apache.commons.math3.primes.Primes;
import org.bouncycastle.jce.provider.BouncyCastleProvider;

import java.security.Security;

public class NumberTheoryExploration {

    public static void main(String[] args) {

        Security.addProvider(new BouncyCastleProvider());

        // TODO 1: Basic integers

        int a = 3, b = 18, n = 25;

        // TODO 1.1: GCD
        System.out.println("gcd(a,b) = " + ArithmeticUtils.gcd(a, b));

        // TODO 1.2: LCM
        System.out.println("lcm(a,b) = " + ArithmeticUtils.lcm(a, b));

        // TODO 1.3: Divisors of n
        System.out.println("Divisors of n:");
        for (int d = 1; d <= n; d++) {
            if (n % d == 0) {
                System.out.print(d + " ");
            }
        }
        System.out.println();

        // TODO 1.4: Number of divisors
        int divisorCount = 0;
        for (int d = 1; d <= n; d++) {
            if (n % d == 0) divisorCount++;
        }
        System.out.println("Number of divisors of n = " + divisorCount);

        // TODO 1.5: Prime factorization
        System.out.println("Prime factorization of n:");
        factorize(n);

        // TODO 2: Primality
        System.out.println("Is n prime? " + Primes.isPrime(n));

        // TODO 3: Euler's Totient Function
        System.out.println("phi(a) = " + totient(a));
        System.out.println("phi(b) = " + totient(b));
        System.out.println("phi(n) = " + totient(n));

        // TODO 4: Chinese Remainder Theorem (CRT)
        BigInteger[] moduli = {
            BigInteger.valueOf(99),
            BigInteger.valueOf(97),
            BigInteger.valueOf(95)
        };

        BigInteger[] residues = {
            BigInteger.valueOf(49),
            BigInteger.valueOf(76),
            BigInteger.valueOf(65)
        };

        BigInteger crtSolution = chineseRemainder(residues, moduli);
        System.out.println("CRT solution x = " + crtSolution);

        // TODO 5: Miller–Rabin Primality Test
        BigInteger test = BigInteger.valueOf(1_373_651);
        System.out.println("Miller–Rabin result: " +
                test.isProbablePrime(10));

        // TODO 6: Primitive Roots (limited implementation)
        System.out.println("Primitive root modulo 25:");
        System.out.println(findPrimitiveRoot(25));

        // TODO 7: Discrete Logarithm (brute-force demo)
        System.out.println("Solving 7^x ≡ 15 (mod 41)");
        System.out.println("x = " + discreteLog(7, 15, 41));
    }

    // ------------------ Helpers ------------------
    static void factorize(int n) {
        int x = n;
        for (int p = 2; p * p <= x; p++) {
            while (x % p == 0) {
                System.out.print(p + " ");
                x /= p;
            }
        }
        if (x > 1) System.out.print(x);
        System.out.println();
    }

    static int totient(int n) {
        int result = n;
        for (int p = 2; p * p <= n; p++) {
            if (n % p == 0) {
                while (n % p == 0) n /= p;
                result -= result / p;
            }
        }
        if (n > 1) result -= result / n;
        return result;
    }

    static BigInteger chineseRemainder(BigInteger[] a, BigInteger[] m) {
        BigInteger M = BigInteger.ONE;
        for (BigInteger mi : m) M = M.multiply(mi);

        BigInteger result = BigInteger.ZERO;
        for (int i = 0; i < m.length; i++) {
            BigInteger Mi = M.divide(m[i]);
            BigInteger inv = Mi.modInverse(m[i]);
            result = result.add(a[i].multiply(Mi).multiply(inv));
        }
        return result.mod(M);
    }

    static int findPrimitiveRoot(int n) {
        if (totient(n) <= 2) return -1;
        int phi = totient(n);

        for (int g = 2; g < n; g++) {
            Set<Integer> seen = new HashSet<>();
            int x = 1;
            for (int i = 0; i < phi; i++) {
                x = (x * g) % n;
                seen.add(x);
            }
            if (seen.size() == phi) return g;
        }
        return -1;
    }

    static int discreteLog(int base, int target, int mod) {
        int value = 1;
        for (int x = 0; x < mod; x++) {
            if (value == target) return x;
            value = (value * base) % mod;
        }
        return -1;
    }
}

File: Chapter_3/numbertheory.py
"""
TODO-based exploration of number theory with SymPy.
Author: Miguel Vargas Martin using SymPy manual and ChatGPT
Last modified: January 2026

Instructions:
- Read each TODO carefully.
- Before running a line, try to predict its output.
- Uncomment lines only when instructed.
- Answer the questions in the comments or in a separate worksheet.
"""

import warnings
warnings.filterwarnings('ignore')

from sympy import *

# TODO 1: Basic integers

a, b, n = 3, 18, 25

# TODO 1.1:
# Predict gcd(a, b). Then uncomment and verify.
# print(igcd(a, b))

# TODO 1.2:
# Predict lcm(a, b). What is the relationship between gcd and lcm?
# print(ilcm(a, b))

# TODO 1.3:
# List all positive divisors of n.
# How does this relate to the prime factorization of n?
# print(divisors(n))

# TODO 1.4:
# How many divisors does n have?
# print(divisor_count(n))

# TODO 1.5:
# Find the prime factorization of n.
# print(factorint(n))


# TODO 2: Primality

# TODO 2.1:
# Is n prime? Why or why not?
# print(isprime(n))

# TODO 2.2:
# Pick a different value for n that *is* prime and test it.


# TODO 3: Euler's Totient Function

from sympy.ntheory import totient

# TODO 3.1:
# Compute φ(a), φ(b), and φ(n).
# What does φ(n) count?
# print(totient(a))
# print(totient(b))
# print(totient(n))

# TODO 3.2:
# Verify φ(p) = p − 1 for a prime p of your choice.


# TODO 4: Chinese Remainder Theorem (CRT)

from sympy.ntheory.modular import crt

# Solve the system:
#   x ≡ 49 (mod 99)
#   x ≡ 76 (mod 97)
#   x ≡ 65 (mod 95)

# TODO 4.1:
# Before running this, check whether the moduli are pairwise coprime.
solution = crt([99, 97, 95], [49, 76, 65])
print(solution)

# TODO 4.2:
# Extract the solution x and verify each congruence manually.


# TODO 5: Miller–Rabin Primality Test

from sympy.ntheory.primetest import mr

# TODO 5.1:
# The number below is large. Do you expect it to be prime?
print(mr(1373651, [2, 3, 5]))

# TODO 5.2:
# Change the bases. Does the result change?
# Why is Miller–Rabin called a "probabilistic" test?


# TODO 6: Primitive Roots

# TODO 6.1:
# Is a a primitive root modulo n?
print(is_primitive_root(a, n))

# TODO 6.2:
# Find a primitive root modulo n (if one exists).
g = primitive_root(n)
print(g)

# TODO 6.3:
# Verify that g generates all units modulo n.


# TODO 7: Discrete Logarithm

from sympy.ntheory import discrete_log

# Solve: find x such that 7^x ≡ 15 (mod 41)
x = discrete_log(41, 15, 7)
print(x)

# TODO 7.1:
# Verify the result by computing 7**x % 41.
# TODO 7.2:
# Why is the discrete logarithm problem considered hard in general?
